packages/contracts/src/contracts/current/protocol/Exchange/MixinSignatureValidator.sol

```diff
@@ -131,20 +128,48 @@ contract MixinSignatureValidator is
             isValid = signer == recovered;
             return isValid;
 
-        // Signature using EIP712
-        } else if (signatureType == SignatureType.EIP712) {
+        // Implicitly signed by signer of Ethereum transaction.
+        // This is useful when initiating a call from a contract that might
+        // otherwise require multiple signatures.
+        // Example: Contract A calls Exchange.fillOrder using `executeTransaction`.
+        // This would normally require the user to sign both an Ethereum transaction
+        // and a 0x transaction. By using the TxOrigin signature type, the signature
+        // for the Ethereum transaction will encompass both signatures.
+        } else if (signatureType == SignatureType.TxOrigin) {
             require(
-                signature.length == 66,
+                signature.length == 1,
                 INVALID_SIGNATURE_LENGTH
             );
-            v = uint8(signature[1]);
-            r = readBytes32(signature, 2);
-            s = readBytes32(signature, 34);
-            recovered = ecrecover(hash, v, r, s);
-            isValid = signer == recovered;
+            isValid = signer == tx.origin;
```

Comment:
`tx.origin` is very tempting for certain problems, but it is filled with security issues. Basically, the moment you use it for transaction verification, you open the door to phishing attacks by other contracts. Solidity devs agree you should not use it for authentication (like this PR aims to do).

In fact, it would have been removed if it wasn't for backwards compatibility:
https://github.com/ethereum/solidity/issues/683

It has also been added as a warning in static analysis:
https://github.com/ethereum/remix-ide/issues/173

IMHO the security risks (and public perception) are not worth the gain in user experience. I also think auditors would and should raise this as an issue.

Can we revisit the problem this is trying to address and come up with a different solution?
Perhaps the `SignatureType.Presigned` can be used. Or we can lobby for batch transactions in Ethereum. Or for batch signing in the UI clients.

Reply:
In my opinion, using `tx.origin` is fine when you use signature in the case of 0x. tx.origin is terrible when it's used as ownership for instance, since that's really easy to fish users for and then call the owned contract. In general,  Early alarms scared people away from this approach, but I think tx.origin can be powerful if all corner cases are taken into consideration.

In the present proposal, a malicious contract would need to 

1. Fool the maker to call their contract
2. Pass in a valid signed order from the maker that specified signature type as `SignatureType.TxOrigin`
3. Use this valid signature for a malicious purpose ...

For 3, the only thing I can see is that the maker order could be filled by the malicious contract, which doesn't seem to bad. What else could happen?

Reply:
There are definitely some risks here (somewhat undecided on if this should make the cut). For example:

- You call some function on a contract that calls a chain of other contracts. Somewhere along the way, a contract calls fillOrder using the TxOrigin signature type, making you unintentionally fill a malicious signed order.
- In a similar vein, the contract could call fillOrder on a malicious _unsigned_ order but pass in a TxOrigin signature for that order. This pretty much has the same result as the previous scenario.

What makes this tricky is that this could be hidden in a long chain of contract calls, making it hard to audit beforehand. If signers like MetaMask showed you how a transaction would change the state, the risk of this happening could decrease a lot.

Reply:
Oh, I see. I just realized that there is no "0x" signature required from tx.origin, apologies. Yeah, this can definitely lead to bad stuff.